polars/polars-lazy/polars-plan/src/logical_plan/optimizer/flatten_union.rs

```diff
@@ -0,0 +1,47 @@
+use polars_utils::arena::{Arena, Node};
+use ALogicalPlan::*;
+
+use super::OptimizationRule;
+use crate::prelude::ALogicalPlan;
+
+pub struct FlattenUnionRule {}
+
+fn get_union_inputs(node: Node, lp_arena: &Arena<ALogicalPlan>) -> Option<&[Node]> {
+    match lp_arena.get(node) {
+        ALogicalPlan::Union { inputs, .. } => Some(inputs),
+        _ => None,
+    }
+}
+
+impl OptimizationRule for FlattenUnionRule {
+    fn optimize_plan(
+        &mut self,
+        lp_arena: &mut polars_utils::arena::Arena<ALogicalPlan>,
+        _expr_arena: &mut polars_utils::arena::Arena<crate::prelude::AExpr>,
+        node: polars_utils::arena::Node,
+    ) -> Option<ALogicalPlan> {
+        let lp = lp_arena.get(node);
+
+        match lp {
+            Union { inputs, options }
+                if inputs
```

Comment:
Can we add a state `flattened_by_opt` to the `UnionOptions`. This will prevent we run this loop every time this rule is checked. It is not uncommon to have unions over a lot of files, so it is worth it to return early in such a case.

Reply:
not sure I 100% follow. If we add a `flattened_by_opt` when would we set it? and when would we return early if it were set?

Reply:
@ritchie46, I added the opt state `flattened_by_opt`. This should be ready for final review.